Added ability for users to forfeit game

jy300x2 opened the pull request:
- Added forfeit, active, and winner_id fields to games table (also added indexes for each)
- Created forfeit! method in game model
- Created forfeit action in games controller
- Added forfeit route to routes
- Added forfeit method tests to games spec 
- Added forfeit button to allow users to forfeit from game
- Showed on game#show page whether game was active or inactive 
- Added CSS to style active vs. inactive game status

mscoutermarsh commented on app/controllers/games_controller.rb:
```diff
@@ -30,6 +30,13 @@ def update
 
+  def forfeit
+    game = Game.find(params[:id])
+    return render_not_found(:unauthorized) if @game.white_player_id.present? && @game.black_player_id.present?
```
Not understanding this line.

They won't be able to forfeit if the game has 2 players?

Also, `@game` here will be nil. Would need to be `game`.

mscoutermarsh commented on app/models/game.rb:
```diff
@@ -5,6 +5,7 @@ class Game < ActiveRecord::Base
   has_many :pieces
   belongs_to :black_player, class_name: 'User'
   belongs_to :white_player, class_name: 'User'
+  belongs_to :winner, class_name: 'User'
```
:+1:

mscoutermarsh commented on spec/models/game_spec.rb:
```diff
@@ -116,4 +116,25 @@
     end
   end
+
+  describe 'end game' do
```
it's good to have the method name in the describe.

Like this

`describe '#forfeit'`

#method_name = "this is an instance method"
.method_name = "this is a class method"

mscoutermarsh commented on app/models/game.rb:
```diff
@@ -61,4 +62,18 @@ def location_is_under_attack_by_color?(color, row, col)
+
+  def forfeit!(user)
+    player_ids = [black_player_id, white_player_id]
+    unique_player_ids = player_ids.uniq
```
If it's currently possible to have the same player as both `black` and `white`. We should add a validation to the model so that this never happens. Then you won't have to worry about this case here.

mscoutermarsh commented on app/models/game.rb:
```diff
@@ -61,4 +62,18 @@ def location_is_under_attack_by_color?(color, row, col)
   end
+
+  def forfeit!(user)
+    player_ids = [black_player_id, white_player_id]
```
instead of working with `ids`. I think it would be easier to use the object directly.

```Ruby
players = [black_player, white_player]
winner = players.find{ |player| player.id != forfeiting_user }
```

mscoutermarsh commented on app/models/game.rb:
```diff
@@ -61,4 +62,18 @@ def location_is_under_attack_by_color?(color, row, col)
     false
   end
+
+  def forfeit!(user)
```
Instead of `user`, could be super explicit here and use something like `forfeiting_user`. Then later in your method, it's super obvious which user lost.

jy300x2 commented:
- Changed name of forfeit! method to forfeiting_user
- Refactored forfeiting_user method 
- Refactored forfeit action in games controller
- Refactored tests for forfeiting_user method
- Refactored game show page so that only the current user can see their “Forfeit” button (and not the other player’s)
